cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/DefaultMetadataHandler.java

```diff
@@ -0,0 +1,91 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metrics.process;
+
+import io.cdap.cdap.proto.id.TopicId;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Default implementation for {@link MetadataHandler}.
+ * Loads/saves metadata from/to {@link MetricsConsumerMetaTable}
+ */
+public class DefaultMetadataHandler implements MetadataHandler {
+
+  private final ConcurrentMap<MetricsMetaKey, TopicProcessMeta> topicProcessMetaMap;
+  private MetricsConsumerMetaTable metaTable;
+  private final String metricsPrefixForDelayMetrics;
+  private final MetricsMetaKeyProvider keyProvider;
+
+  public DefaultMetadataHandler(String metricsPrefixForDelayMetrics, MetricsMetaKeyProvider keyProvider) {
+    this.topicProcessMetaMap = new ConcurrentHashMap<>();
+    this.metricsPrefixForDelayMetrics = metricsPrefixForDelayMetrics;
+    this.keyProvider = keyProvider;
+  }
+
+  @Override
+  public void initCache(List<TopicId> metricsTopics, MetricsConsumerMetaTable metaTable) {
+    this.metaTable = metaTable;
+    Map<TopicId, MetricsMetaKey> keyMap = keyProvider.getKeys(metricsTopics);
+    for (Map.Entry<TopicId, MetricsMetaKey> keyEntry : keyMap.entrySet()) {
+      MetricsMetaKey topicIdMetaKey = keyEntry.getValue();
+      TopicProcessMeta topicProcessMeta = metaTable.getTopicProcessMeta(topicIdMetaKey);
+      if (topicProcessMeta == null || topicProcessMeta.getMessageId() == null) {
+        continue;
+      }
+      String oldestTsMetricName = String.format("%s.topic.%s.oldest.delay.ms",
+                                                metricsPrefixForDelayMetrics, keyEntry.getKey());
+      String latestTsMetricName = String.format("%s.topic.%s.latest.delay.ms",
+                                                metricsPrefixForDelayMetrics, keyEntry.getKey());
+      topicProcessMetaMap.put(topicIdMetaKey,
+                              new TopicProcessMeta(topicProcessMeta.getMessageId(),
+                                                   topicProcessMeta.getOldestMetricsTimestamp(),
+                                                   topicProcessMeta.getLatestMetricsTimestamp(),
+                                                   topicProcessMeta.getMessagesProcessed(),
+                                                   topicProcessMeta.getLastProcessedTimestamp(),
+                                                   oldestTsMetricName, latestTsMetricName));
+    }
+  }
+
+  @Override
+  public void updateCache(MetricsMetaKey key, TopicProcessMeta value) {
+    topicProcessMetaMap.put(key, value);
+  }
+
+  @Override
+  public void saveCache(Map<MetricsMetaKey, TopicProcessMeta> topicProcessMetaMap) {
+    if (topicProcessMetaMap.isEmpty()) {
+      return;
+    }
+    metaTable.saveMetricsProcessorStats(topicProcessMetaMap);
```

Comment:
In persistTopicProcessMeta(), we catch Exception. Should we do the same here? Also, can we keep the comment that explains when topicProcessMetaMap can be empty?

Reply:
This is called by persistMetricsAndTopicProcessMeta() which handles the exception. Added the comment.